### PR TITLE
Allow encoded HTTP URL to create new entry 

### PR DESCRIPTION
**Pull request by KonstantinosChanioglou**
## Issue Link
This pull request resolves partially #10648 (Entry creation from DOI does not recognize URL encoding, but "get bibliographical data" does.)

## Issue description
Error is occured when the user enters encoded HTTP URL in the DOI field in order create a new entry. The error occurs only in the case of entering an HTTP URL with both ways (new entry/get bibliographical data). When the DOI only is given the application creates the entry normally with both ways.

## Bug Cause
The cause of the error was the creation of new URI (URI url = new URI(trimmedDoi)) which was throwing `java.lang.IllegalArgumentException: java.net.URISyntaxException: Illegal character in path at index 42: https://doi.org/10.1175/1520-0493(2002)130<1913:EDAWPO>2.0.CO;2 `.

## Modifications and Results
At `org.jabref.model.entry.identifier.DOI#DOI`, the method with which the trimmedDoi is decoded has been modified in the case trimmedDoi being an HTTP URL.

After the modification the encoded HTTP URLs are decoded correctly and the entry is created correctly

## Tests Added
In the  `org.jabref.model.entry.identifier.DOITest` the next tests have been added:

1.  `acceptEncodedDoiUrl`
Encoded HTTP URL is passed as a parameter and needs to be decoded and the DOI to be extracted correctly.
3. `rejectNullDoiParameter`
Test case for giving `Null` for the DOI. There is no case that the vallue of `trimmedDoi` become `Null` other that doi parameter is null. 

Note: For the case of the given doi parameter is not HTTP URL the modifications for this issue are not relevant and the already existed functionality handles them correctly.

### CODE REVIEW

File: src/test/java/org/jabref/model/entry/identifier/DOITest.java [hunk 1]
```diff
@@ -306,4 +306,15 @@ public void rejectMissingDivider() {
     public void rejectMissingDividerInShortDoi() {
         assertThrows(IllegalArgumentException.class, () -> new DOI("10gf4gqc end"));
     }
+
+    @Test
+    public void acceptEncodedDoiUrl() {
+        assertEquals("10.1175/1520-0493(2002)130<1913:EDAWPO>2.0.CO;2", new DOI("https://doi.org/10.1175/1520-0493(2002)130<1913:EDAWPO>2.0.CO;2").getDOI());
+        assertEquals("10.1175/1520-0493(2002)130<1913:EDAWPO>2.0.CO;2", new DOI("https://doi.org/10.1175/1520-0493(2002)130%3C1913:EDAWPO%3E2.0.CO;2").getDOI());
```
Comment (koppor) [inline src/test/java/org/jabref/model/entry/identifier/DOITest.java:313]: This should go to a parameterized test, `@CsvSource` (https://mikemybytes.com/2021/10/19/parameterize-like-a-pro-with-junit-5-csvsource/)

**Review by koppor — CHANGES_REQUESTED** [review]
Nice find! And thank you for contributing!

Do you have time to fix the issues reported by the CI? (See the red crosses at the end of your PR)

- /DOI.java:130: Prefer java.net.URLDecoder.decode(String, Charset)
- `ShortenFullyQualifiedTypeReferences` (can be fixed by running gradle's `rewriteRun` task)

File: src/test/java/org/jabref/model/entry/identifier/DOITest.java (hunk 1 above)
Comment (koppor) [reply src/test/java/org/jabref/model/entry/identifier/DOITest.java:313]: Oh, we do have `testData()` very up in the class. Please use that!

Comment (KonstantinosChanioglou) [reply src/test/java/org/jabref/model/entry/identifier/DOITest.java]: Thank you very much for your feedback!

I have now pushed my changes concerning:

1. /DOI.java:130: Prefer `java.net.URLDecoder.decode(String, Charset)`
2. Moving suggested tests in `testData()`

In addition I have executed `rewriteRun` task from gradle

File: src/test/java/org/jabref/model/entry/identifier/DOITest.java [hunk 2]
```diff
@@ -161,6 +161,11 @@ private static Stream<Arguments> testData() {
                 // ? -> (%3F)
                 Arguments.of("https://doi.org/10.1006/rwei.1999%3F.0001",
                         new DOI("https://doi.org/10.1006/rwei.1999%3F.0001").getURIAsASCIIString()),
+                Arguments.of("10.1175/1520-0493(2002)130<1913:EDAWPO>2.0.CO;2", new DOI("https://doi.org/10.1175/1520-0493(2002)130%3C1913:EDAWPO%3E2.0.CO;2").getDOI()),
+                // <,> -> (%3C, %3E)
```
Comment (koppor) [inline src/test/java/org/jabref/model/entry/identifier/DOITest.java:165]: Why the comment below the check?

Comment (KonstantinosChanioglou) [reply src/test/java/org/jabref/model/entry/identifier/DOITest.java]: Excuse me for that, I have fixed it.

**Comment by koppor** [comment]
I tried out. When trying with the "new entry from id" tool, I get 404.

image

**Comment by KonstantinosChanioglou** [comment]
> I tried out. When trying with the "new entry from id" tool, I get 404.
> 
> image


From the definition of this issue the changes have to do with the next ways of adding new entry:

1. Using "New entry" option:
New entry

2. Using "Get bibliographical data" option by creating first an empty entry/article:
new article

However I did an analysis for  the "Import by ID" option just to be sure and I found the next points:

1. The new code implementation is not executed while the parameter to the DOI constructor is not an HTTP URL
Import
DOI

2. I tested this option in the (original) application without inluding my changes and it does not work as well (404). So I guess this could be handled in a new github issue.

In addition I observed that in both original application and the one with my changes after creating the entry with the HTTP URL or with only using the DOI (in the original application) the tab "Citation relations" contains an error which is not related with the new implementation of this issue and possibly can be addressed in a new issue:
Citation relations

**Comment by koppor** [comment]
> The new code implementation is not executed while the parameter to the DOI constructor is not an HTTP URL

The screenshot shows a http URL, therefore I updated the PR description that this PR partially fixes it. I also commented on the issue that entry-from-id is still open.

**Review by Siedlerchr — APPROVED** [review]

**Comment by Siedlerchr** [comment]
Thanks, we will do a follow up issue with the other fetcher parsign

**Pull request merged**